drivers/frequency/adf5355/adf5355.c

```diff
@@ -0,0 +1,518 @@
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <inttypes.h>
+#include <stdio.h>
+#include "error.h"
+#include <malloc.h>
+#include "delay.h"
+#include "util.h"
+#include "adf5355.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * SPI register write to device.
+ * @param dev - The device structure.
+ * @param reg - The register address.
+ * @param val - The register data.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t adf5355_write(struct adf5355_dev *dev,
+			     uint8_t reg_addr,
+			     uint32_t data)
+{
+	uint8_t buf[ADF5355_SPI_NO_BYTES];
+	data = data | reg_addr;
```

Comment:
This seems a bit odd. Is `data` supposed to be in a specific format that (i think shifted by 8 bits in this case)? If yes, it would be good to document it, but it might not be necessary if you just set do:
`buff[0] = data >> 16;
 buff[1] = data >> 8;
 buff[2] = data;
 buff[3] = reg_addr;`

Reply:
the reg_addr is only 4 bits, the rest of 4 are data bytes.
I know this is a strange approach, but I used it because there were remarks on other similar drivers that the read/write function signature should comply to the IIO support.
Same thing happens here:
https://github.com/analogdevicesinc/no-OS/blob/ca092f47e8ed60e7fb94b265fdcf27da12b6c3d9/drivers/frequency/adf5902/adf5902.c#L64

Reply:
Understood. So is there any specific documentation about the format of the parameters, for example data or register address needing to be shifted prior to passing them to the function? Assuming the user does not use the defines in the header.

Reply:
The defines (register bits masks) are handling this data shifting. So if the customer will use our driver as it is, there shouldn't be any problems.
But I agree on documenting the data/register address handling. But not sure where it should be done. Maybe in the function header comment?

Reply:
Added a note in the function header comment.